File: testdata/composite.js










var PASS = true;

function person(name, age) {
	this.name=name;
	this.age=age
}


function older(p1, p2) {
	if (p1.age > p2.age) {
		return [p1, p1.age - p2.age];
	}
	return [p2, p2.age - p1.age];
}


function older10(people) {
	var older = people.v[0];


	for (var index = 1; index < 10; index++) {
		if (people.v[index].age > older.age) {
			older = people.v[index];
		}
	}
	return older;
}




function builtInArray() {
	var pass = true;





	var a1 = g.MkArray([5], 0);
	var a2 = g.MkArray([5], 0);
	var a3 = g.MkArray([5], 0, [2]);
	var a4 = g.MkArray([5], 0, [2, 4]);

	var a5 = g.MkArray([3,4], 0);
	var a6 = g.MkArray([3,4,2], 0);

	var _ = function(msg, in_, out) { return {
		msg: msg,
		in_: in_,
		out: out
	};}; var tests = [


		_("len a1", a1.len() == 5, true),
		_("len a2", a2.len() == 5, true),
		_("len a3", a3.len() == 5, true),
		_("len a4", a4.len() == 5, true),
		_("len a4", a4.len() != 5, false),

		_("cap a1", a1.cap() == 5, true),
		_("cap a2", a2.cap() == 5, true),
		_("cap a3", a3.cap() == 5, true),
		_("cap a4", a4.cap() == 5, true),

		_("len a5", a5.len() == 3, true),
		_("cap a5", a5.cap() == 3, true),
		_("len a5[0]", a5.len(0) == 4, true),
		_("cap a5[0]", a5.cap(0) == 4, true),
		_("len a5[1000]", a5.len(1000) == 4, true),
		_("cap a5[1000]", a5.cap(1000) == 4, true),

		_("len a6", a6.len() == 3, true),
		_("cap a6", a6.cap() == 3, true),
		_("len a6[0]", a6.len(0) == 4, true),
		_("cap a6[0]", a6.cap(0) == 4, true),
		_("len a6[0][0]", a6.len(0,0) == 2, true),
		_("cap a6[0][0]", a6.cap(0,0) == 2, true),
		_("len a6[0][1000]", a6.len(0,1000) == 2, true),
		_("cap a6[0][1000]", a6.cap(0,1000) == 2, true)
	];

	var t; for (var _ in tests) { t = tests[_];
		if (JSON.stringify(t.in_) != JSON.stringify(t.out)) {
			document.write("&nbsp;&nbsp;&nbsp;&nbsp;FAIL: " + t.msg + " => got " + t.in_ + ", want " + t.out + "<br>");
			pass = false, PASS = false;
		}
	}
	if (pass) {
		document.write("&nbsp;&nbsp;&nbsp;&nbsp;pass<br>");
	}
}

function initArray() {
	var pass = true;


	var array1 = g.MkArray([10], new person("", 0), [
		new person("", 0),
		new person("Paul", 23),
		new person("Jim", 24),
		new person("Sam", 84),
		new person("Rob", 54),
		new person("", 0),
		new person("", 0),
		new person("", 0),
		new person("Karl", 10),
		new person("", 0)
	]);


	var array2 = g.MkArray([10], new person("", 0), [
		new person("", 0),
		new person("Paul", 23),
		new person("Jim", 24),
		new person("Sam", 84),
		new person("Rob", 54),
		new person("", 0),
		new person("", 0),
		new person("", 0),
		new person("Karl", 10),
		new person("", 0)]);

	var _ = function(msg, in_, out) { return {
		msg: msg,
		in_: in_,
		out: out
	};}; var tests = [
		_("len", array1.len() == array2.len(), true),
		_("cap", array1.cap() == array2.cap(), true),
		_("equality", JSON.stringify(array1.v) == JSON.stringify(array2.v), true)
	];

	var t; for (var _ in tests) { t = tests[_];
		if (JSON.stringify(t.in_) != JSON.stringify(t.out)) {
			document.write("&nbsp;&nbsp;&nbsp;&nbsp;FAIL: " + t.msg + " => got " + t.in_ + ", want " + t.out + "<br>");
			pass = false, PASS = false;
		}
	}
	if (pass) {
		document.write("&nbsp;&nbsp;&nbsp;&nbsp;pass<br>");
	}
}

function _array() {

	var array = g.MkArray([10], new person("", 0));



	array.v[1] = new person("Paul", 23);
	array.v[2] = new person("Jim", 24);
	array.v[3] = new person("Sam", 84);
	array.v[4] = new person("Rob", 54);
	array.v[8] = new person("Karl", 19);

	var older = older10(array);

	if (older.name == "Sam") {
		document.write("&nbsp;&nbsp;&nbsp;&nbsp;pass<br>");
	} else {
		document.write("&nbsp;&nbsp;&nbsp;&nbsp;FAIL: got " + older.name + ", want Sam<br>");
		PASS = false;
	}
}

function multiArray() {

	var doubleArray_1 = g.MkArray([2,4], 0, [[1, 2, 3, 4], [5, 6, 7, 8]]);


	var doubleArray_2 = g.MkArray([2,4], 0, [
		[1, 2, 3, 4], [5, 6, 7, 8]]);


	var doubleArray_3 = g.MkArray([2,4], 0, [
		[1, 2, 3, 4],
		[5, 6, 7, 8]
	]);

	if (JSON.stringify(doubleArray_1.v) == JSON.stringify(doubleArray_2.v) && JSON.stringify(doubleArray_2.v) == JSON.stringify(doubleArray_3.v)) {
		document.write("&nbsp;&nbsp;&nbsp;&nbsp;pass<br>");
	} else {
		document.write("&nbsp;&nbsp;&nbsp;&nbsp;FAIL: got different arraies<br>");
		PASS = false;
	}
}




function _struct() {
	var pass = true;

	var tom = new person("", 0);
	tom.name = "Tom", tom.age = 18;

	var bob = new person(); bob.age = 25, bob.name = "Bob";
	var paul = new person("Paul", 43);

	var _ = older(tom, bob), TB_older = _[0], TB_diff = _[1];
	var _ = older(tom, paul), TP_older = _[0], TP_diff = _[1];
	var _ = older(bob, paul), BP_older = _[0], BP_diff = _[1];

	var _ = function(msg, inPerson, outPerson, inDiff, outDiff) { return {
		msg: msg,
		inPerson: inPerson,
		outPerson: outPerson,
		inDiff: inDiff,
		outDiff: outDiff
	};}; var tests = [
		_("Tom,Bob", TB_older, bob, TB_diff, 7),
		_("Tom,Paul", TP_older, paul, TP_diff, 25),
		_("Bob,Paul", BP_older, paul, BP_diff, 18)
	];

	var t; for (var _ in tests) { t = tests[_];
		if (JSON.stringify(t.inPerson) != JSON.stringify(t.outPerson)) {
			document.write("&nbsp;&nbsp;&nbsp;&nbsp;FAIL: " + t.msg + " => person got " + t.inPerson + ", want " + t.outPerson + "<br>");

			pass = false, PASS = false;
		}
		if (JSON.stringify(t.inDiff) != JSON.stringify(t.outDiff)) {
			document.write("&nbsp;&nbsp;&nbsp;&nbsp;FAIL: " + t.msg + " => difference got " + t.inDiff + ", want " + t.outDiff + "<br>");

			pass = false, PASS = false;
		}
	}
	if (pass) {
		document.write("&nbsp;&nbsp;&nbsp;&nbsp;pass<br>");
	}
}

function main() {
	document.write("<br><br>== Composite types<br><br>");

	document.write("=== RUN builtInArray<br>");
	builtInArray();
	document.write("=== RUN initArray<br>");
	initArray();
	document.write("=== RUN array<br>");
	_array();
	document.write("=== RUN multiArray<br>");
	multiArray();

	document.write("=== RUN struct<br>");
	_struct();

	if (PASS) {
		document.write("PASS<br>");
	} else {
		document.write("FAIL<br>");
		alert("Fail: Composite types");
	}
} main();
/* Generated by Go2js (github.com/tredoe/go2js) */

File: testdata/composite.go
// Copyright 2011 Jonas mg
//
// This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
// If a copy of the MPL was not distributed with this file, You can obtain one at
// http://mozilla.org/MPL/2.0/.

package main

import "fmt"

var PASS = true

type person struct {
	name string
	age  int
}

// Return the older person of p1 and p2, and the difference in their ages.
func older(p1, p2 person) (person, int) {
	if p1.age > p2.age {
		return p1, p1.age - p2.age
	}
	return p2, p2.age - p1.age
}

// Return the older person in a group of 10 persons.
func older10(people [10]person) person {
	older := people[0] // The first one is the older for now.

	// Loop through the array and check if we could find an older person.
	for index := 1; index < 10; index++ { // We skipped the first element here.
		if people[index].age > older.age {
			older = people[index]
		}
	}
	return older
}

// == Array
//

func builtInArray() {
	pass := true

	// TODO
	//var pa1 *[4]int
	//pa2 := new([4]int) // *[4]int

	var a1 [5]int
	a2 := [5]int{}
	a3 := [5]int{2}
	a4 := [5]int{2, 4}

	a5 := [3][4]int{}
	a6 := [3][4][2]int{}

	tests := []struct {
		msg string
		in  bool
		out bool
	}{
		//{"nil a1", a1 == nil, true},
		//{"nil a2", a2 == nil, false},
		{"len a1", len(a1) == 5, true},
		{"len a2", len(a2) == 5, true},
		{"len a3", len(a3) == 5, true},
		{"len a4", len(a4) == 5, true},
		{"len a4", len(a4) != 5, false},

		{"cap a1", cap(a1) == 5, true},
		{"cap a2", cap(a2) == 5, true},
		{"cap a3", cap(a3) == 5, true},
		{"cap a4", cap(a4) == 5, true},

		{"len a5", len(a5) == 3, true},
		{"cap a5", cap(a5) == 3, true},
		{"len a5[0]", len(a5[0]) == 4, true},
		{"cap a5[0]", cap(a5[0]) == 4, true},
		{"len a5[1000]", len(a5[1000]) == 4, true},
		{"cap a5[1000]", cap(a5[1000]) == 4, true},

		{"len a6", len(a6) == 3, true},
		{"cap a6", cap(a6) == 3, true},
		{"len a6[0]", len(a6[0]) == 4, true},
		{"cap a6[0]", cap(a6[0]) == 4, true},
		{"len a6[0][0]", len(a6[0][0]) == 2, true},
		{"cap a6[0][0]", cap(a6[0][0]) == 2, true},
		{"len a6[0][1000]", len(a6[0][1000]) == 2, true},
		{"cap a6[0][1000]", cap(a6[0][1000]) == 2, true},
	}

	for _, t := range tests {
		if t.in != t.out {
			fmt.Printf("\tFAIL: %s => got %v, want %v\n", t.msg, t.in, t.out)
			pass, PASS = false, false
		}
	}
	if pass {
		fmt.Println("\tpass")
	}
}

func initArray() {
	pass := true

	// Declare and initialize an array A of 10 person.
	array1 := [10]person{
		person{"", 0},
		person{"Paul", 23},
		person{"Jim", 24},
		person{"Sam", 84},
		person{"Rob", 54},
		person{"", 0},
		person{"", 0},
		person{"", 0},
		person{"Karl", 10},
		person{"", 0},
	}

	// Declare and initialize an array of 10 persons, but let the compiler guess the size.
	array2 := [...]person{ // Substitute '...' instead of an integer size.
		person{"", 0},
		person{"Paul", 23},
		person{"Jim", 24},
		person{"Sam", 84},
		person{"Rob", 54},
		person{"", 0},
		person{"", 0},
		person{"", 0},
		person{"Karl", 10},
		person{"", 0}}

	tests := []struct {
		msg string
		in  bool
		out bool
	}{
		{"len", len(array1) == len(array2), true},
		{"cap", cap(array1) == cap(array2), true},
		{"equality", array1 == array2, true},
	}

	for _, t := range tests {
		if t.in != t.out {
			fmt.Printf("\tFAIL: %s => got %v, want %v\n", t.msg, t.in, t.out)
			pass, PASS = false, false
		}
	}
	if pass {
		fmt.Println("\tpass")
	}
}

func _array() {
	// Declare an example array variable of 10 person called 'array'.
	var array [10]person

	// Initialize some of the elements of the array, the others are by default
	// set to person{"", 0}
	array[1] = person{"Paul", 23}
	array[2] = person{"Jim", 24}
	array[3] = person{"Sam", 84}
	array[4] = person{"Rob", 54}
	array[8] = person{"Karl", 19}

	older := older10(array) // Call the function by passing it our array.

	if older.name == "Sam" {
		fmt.Println("\tpass")
	} else {
		fmt.Printf("\tFAIL: got %v, want Sam\n", older.name)
		PASS = false
	}
}

func multiArray() {
	// Declare and initialize an array of 2 arrays of 4 ints
	doubleArray_1 := [2][4]int{[4]int{1, 2, 3, 4}, [4]int{5, 6, 7, 8}}

	// Simplify the previous declaration, with the '...' syntax
	doubleArray_2 := [2][4]int{
		[...]int{1, 2, 3, 4}, [...]int{5, 6, 7, 8}}

	// Super simpification!
	doubleArray_3 := [2][4]int{
		{1, 2, 3, 4},
		{5, 6, 7, 8},
	}

	if doubleArray_1 == doubleArray_2 && doubleArray_2 == doubleArray_3 {
		fmt.Println("\tpass")
	} else {
		fmt.Print("\tFAIL: got different arraies\n")
		PASS = false
	}
}

// == Struct
//

func _struct() {
	pass := true

	var tom person
	tom.name, tom.age = "Tom", 18

	bob := person{age: 25, name: "Bob"} // specify the fields and their values
	paul := person{"Paul", 43}          // specify values of fields in their order

	TB_older, TB_diff := older(tom, bob)
	TP_older, TP_diff := older(tom, paul)
	BP_older, BP_diff := older(bob, paul)

	tests := []struct {
		msg       string
		inPerson  person
		outPerson person
		inDiff    int
		outDiff   int
	}{
		{"Tom,Bob", TB_older, bob, TB_diff, 7},
		{"Tom,Paul", TP_older, paul, TP_diff, 25},
		{"Bob,Paul", BP_older, paul, BP_diff, 18},
	}

	for _, t := range tests {
		if t.inPerson != t.outPerson {
			fmt.Printf("\tFAIL: %s => person got %v, want %v\n",
				t.msg, t.inPerson, t.outPerson)
			pass, PASS = false, false
		}
		if t.inDiff != t.outDiff {
			fmt.Printf("\tFAIL: %s => difference got %v, want %v\n",
				t.msg, t.inDiff, t.outDiff)
			pass, PASS = false, false
		}
	}
	if pass {
		fmt.Println("\tpass")
	}
}

func main() {
	fmt.Print("\n\n== Composite types\n\n")

	fmt.Println("=== RUN builtInArray")
	builtInArray()
	fmt.Println("=== RUN initArray")
	initArray()
	fmt.Println("=== RUN array")
	_array()
	fmt.Println("=== RUN multiArray")
	multiArray()

	fmt.Println("=== RUN struct")
	_struct()

	if PASS {
		fmt.Println("PASS")
	} else {
		fmt.Println("FAIL")
		print("Fail: Composite types")
	}
}
